src/integrations/kafka/consumer.cpp

```diff
@@ -339,7 +345,22 @@ void Consumer::StartConsuming() {
 
       try {
         consumer_function_(batch);
-        if (const auto err = consumer_->commitSync(); err != RdKafka::ERR_NO_ERROR) {
+        std::vector<RdKafka::TopicPartition *> partitions;
+        utils::OnScopeExit clear_partitions([&]() { RdKafka::TopicPartition::destroy(partitions); });
+
+        if (const auto err = consumer_->assignment(partitions); err != RdKafka::ERR_NO_ERROR) {
+          spdlog::warn("Saving the commited offset of consumer {} failed: {}", info_.consumer_name,
+                       RdKafka::err2str(err));
+          throw ConsumerCheckFailedException(
+              info_.consumer_name,
+              fmt::format("Couldn't get assignment to commit offsets: '{}'", RdKafka::err2str(err)));
+        }
+
+        for (const auto offset = batch.back().Offset() + 1; auto *partition : partitions) {
+          partition->set_offset(offset);
+        }
```

Comment:
I think I suggested this method, but right now it struck me that I was wrong. The delivery order of message only guaranteed inside one partition, but not across partitions. So lets assume we have two partitions:
 - partition#1 with messages `1,3,5`
 - partition#2 with messages `2,4,6`
 
 if our batch size is 3, then it is possible that in the first batch we receive only one of the partitions. In case we receive messages `2,4,6`, we will end up totally ignoring messages from partitions#1.
 
 Thinking about it, I think the `assignment` should return the correct offsets, but we have to verify this first. If that's true, then we don't need to set the offset here explicitly.
 
 @antonio2368 @jbajic @kostasrim do you remember why do we decided to set the offset here?

Reply:
Not sure, my theory was that we need to manually set it after RebalanceCb sets the offset because the internal structure goes out of sync. That means we cannot use `commitSync` without arguments but I'm not sure if that affects the internal partitions and their assignment. :thinking: It would be best if @kostasrim printed out the values of partition offsets without setting them and seeing if it works correctly.

Reply:
Agree. I will check both and update on this. Good catch Benjo!